### ISSUE TITLE
tf.nn.ctc_loss is more numerically unstable when providing sparse instead of dense labels

### ISSUE THREAD

**Issue by f90** [tensorflow/tensorflow#55405]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS BigSur
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.8
- Python version: 3.7

**Describe the current behavior**

`tf.nn.ctc_loss` can return `Inf` value for an input example if the labels are given as sparse tensor, while it computes the value correctly for the same inputs when the labels are given as dense tensor. So it appears that the implementation used when providing dense labels is more numerically stable - this occurs only once the logits coming from the model are more "extreme", as in, very large or very small. However, this issue can and was encountered by me during training an actual model, which broke training.

See example below.

**Describe the expected behavior**

The loss when providing sparse labels is not `Inf` but rather the same value that you get when providing dense labels.

If that cannot be achieved, we should offer a fallback mechanism as shown in my example code, where the dense codepath is used whenever an Inf is encountered.

At the bare minimum, the docs for `tf.nn.ctc_loss` should warn about the numerical instability of the sparse version.

**Standalone code to reproduce the issue**

```
import tensorflow as tf

labels = tf.ones((1, 50), dtype=tf.int64)
logits = tf.concat([tf.ones((1, 1000, 1)), -100 * tf.ones((1, 1000, 1))], axis=2)
label_length = tf.constant([50])
logit_length = tf.constant([1000])

dense_ctc_losses = tf.nn.ctc_loss(
    labels=labels,
    logits=logits,
    label_length=label_length,
    logit_length=logit_length,
    logits_time_major=False,
    blank_index=0,
)

sparse_ctc_losses = tf.nn.ctc_loss(
    labels=tf.sparse.from_dense(tf.cast(tf.convert_to_tensor(labels), dtype=tf.int32)),
    logits=logits,
    label_length=label_length,
    logit_length=tf.cast(logit_length, tf.int32),
    logits_time_major=False,
    blank_index=0,
)

ctc_losses = tf.cond(
    tf.math.reduce_all(tf.math.is_finite(sparse_ctc_losses)),
    lambda: sparse_ctc_losses,
    lambda: tf.nn.ctc_loss(
        labels=labels,
        logits=logits,
        label_length=label_length,
        logit_length=logit_length,
        logits_time_major=False,
        blank_index=0,
    ),
)

# This works fine and produces a proper value
print(f"Dense labels: {dense_ctc_losses}")
# Sparse version gives us an inf value
print(f"Sparse labels: {sparse_ctc_losses}")
# Uses sparse, falls back to dense if sparse gives Inf
print(f"Fallback CTC: {ctc_losses}")
```

Output:
```
2022-03-28 14:35:13.481252: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
WARNING:tensorflow:From /Users/dstoller/.pyenv/versions/lyric-align-baseline/lib/python3.7/site-packages/tensorflow/python/ops/ctc_ops.py:1443: alias_inplace_add (from tensorflow.python.ops.inplace_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Prefer tf.tensor_scatter_nd_add, which offers the same functionality with well-defined read-write semantics.
WARNING:tensorflow:From /Users/dstoller/.pyenv/versions/lyric-align-baseline/lib/python3.7/site-packages/tensorflow/python/ops/ctc_ops.py:1426: alias_inplace_update (from tensorflow.python.ops.inplace_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Prefer tf.tensor_scatter_nd_update, which offers the same functionality with well-defined read-write semantics.
2022-03-28 14:35:15.541318: W ./tensorflow/core/util/ctc/ctc_loss_calculator.h:499] No valid path found.
Dense labels: [706.8936]
Sparse labels: [inf]
Combined CTC: [706.8936]
```

**Comment by f90** [comment]
To add on to this, this is a serious issue because the sparse label version currently runs 50x faster on GPU than the dense label version, but due to the above issue, cannot be used

**Comment by gadagashwini** [comment]
Looks like issue with `SparseTensor support int32`. 
Generally, we see this behaviour because when we are passing a very small number into a function. I could able to reproduce the issue with Tf v2.8 [gist](https://colab.sandbox.google.com/gist/sushreebarsa/5253243ce6846a9b22215a120182b647/55405.ipynb#scrollTo=SVSCfrOLP6W_). Thanks!

**Comment by tilakrayal** [comment]
@f90,
I tried to execute the mentioned code on tf-nightly(2.15.0-dev20230918) and observed that the output value for **Sparse labels** is not INF as earlier versions and it has been changed. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/2c2c398bac4b93505ad6235765d4892d/untitled1411.ipynb). Thank you!
